### PR TITLE
Add functionalities to manipulate content

### PR DESCRIPTION
**Pull request by DenisaCG**
Added backend implementation and connection to frontend contents manager for the following functionalities: 
- create new object
- save existing object with new contents
- rename object
- copy object
- delete object
- check existence of object within drive
- count appearances of object within drive.

For the saving and creation functionalities, a formatting of the content was also needed, depending on its format.

### CODE REVIEW

File: jupyter_drives/manager.py [hunk 1]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
         return response
     
-    async def new_file(self, drive_name, path, **kwargs):
+    async def new_file(self, drive_name, path):
         """Create a new file or directory at the given path.
         
         Args:
             drive_name: name of drive where the new content is created
             path: path where new content should be created
+            content: content of object
```
Comment (afshin) [inline jupyter_drives/manager.py:250]:
```suggestion
```

File: jupyter_drives/manager.py [hunk 2]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+        """
+        data = {}
+        try: 
+            # eliminate leading and trailing backslashes
+            path = path.strip('/')
+
+            if options_format == 'json':
+                formattedContent = json.dumps(content, indent=2)
+                formattedContent = formattedContent.encode("utf-8")
```
Comment (afshin) [inline jupyter_drives/manager.py:293]:
```suggestion
                formatted_content = json.dumps(content, indent=2)
                formatted_content = formatted_content.encode("utf-8")
```

File: jupyter_drives/manager.py [hunk 3]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+                
+                byte_arrays = []
+                for offset in range(0, len(byte_characters), 512):
+                    slice_ = byte_characters[offset:offset + 512]
+                    byte_array = bytearray(slice_)
+                    byte_arrays.append(byte_array)
+                
+                # combine byte arrays and wrap in a BytesIO object 
+                formattedContent = BytesIO(b"".join(byte_arrays))
+                formattedContent.seek(0)  # reset cursor for any further reading
```
Comment (afshin) [inline jupyter_drives/manager.py:306]:
```suggestion
                formatted_content = BytesIO(b"".join(byte_arrays))
                formatted_content.seek(0)  # reset cursor for any further reading
```

File: jupyter_drives/manager.py [hunk 4]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+        }
+        return response
+
+    async def save_file(self, drive_name, path, content, options_format, content_format, content_type):
+        """Save file with new content.
+        
+        Args:
+            drive_name: name of drive where file exists
+            path: path where new content should be saved
+            content: content of object
```
Comment (afshin) [inline jupyter_drives/manager.py:284]: Do you want `options_format`, `content_format`, and `content_type` to be hidden or should they also be in the docstring?

File: jupyter_drives/manager.py [hunk 5]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+                for offset in range(0, len(byte_characters), 512):
+                    slice_ = byte_characters[offset:offset + 512]
+                    byte_array = bytearray(slice_)
+                    byte_arrays.append(byte_array)
+                
+                # combine byte arrays and wrap in a BytesIO object 
+                formattedContent = BytesIO(b"".join(byte_arrays))
+                formattedContent.seek(0)  # reset cursor for any further reading
+            elif options_format == 'text':
+                formattedContent = content.encode("utf-8")
```
Comment (afshin) [inline jupyter_drives/manager.py:308]:
```suggestion
                formatted_content = content.encode("utf-8")
```

File: jupyter_drives/manager.py [hunk 6]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+            else:
+                formattedContent = content
```
Comment (afshin) [inline jupyter_drives/manager.py:310]:
```suggestion
                formatted_content = content
```

File: jupyter_drives/manager.py [hunk 7]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+
+            await obs.put_async(self._content_managers[drive_name], path, formattedContent, mode = "overwrite")
```
Comment (afshin) [inline jupyter_drives/manager.py:312]:
```suggestion
            await obs.put_async(self._content_managers[drive_name], path, formatted_content, mode = "overwrite")
```

File: jupyter_drives/manager.py [hunk 8]
```diff
@@ -237,23 +241,195 @@ async def get_contents(self, drive_name, path):
...
+            }
+        return response
     
-    async def rename_file(self, drive_name, path, **kwargs):
+    async def rename_file(self, drive_name, path, new_path):
         """Rename a file.
         
         Args:
             drive_name: name of drive where file is located
             path: path of file
```
Comment (afshin) [inline jupyter_drives/manager.py:337]:
```suggestion
            path: path of file
            new_path: path of new file name
```

**Review by afshin — APPROVED** [review]
Thank you! Let's merge this when you are ready so that we can get more users on it for testing. Thanks @DenisaCG!

File: jupyter_drives/manager.py (hunk 4 above)
Comment (DenisaCG) [reply jupyter_drives/manager.py:283]: Thanks for catching this! I pushed a commit to add them to the docstring.

**Comment by DenisaCG** [comment]
Great, I'll go ahead and merge it. Thanks for the review!

**Pull request merged**